src/librustc_data_structures/small_c_str.rs

```diff
@@ -11,69 +11,61 @@
 use std::ffi;
 use std::ops::Deref;
 
-const SIZE: usize = 38;
+use smallvec::SmallVec;
+
+const SIZE: usize = 36;
 
 /// Like SmallVec but for C strings.
 #[derive(Clone)]
-pub enum SmallCStr {
-    OnStack {
-        data: [u8; SIZE],
-        len_with_nul: u8,
-    },
-    OnHeap {
-        data: ffi::CString,
-    }
+pub struct SmallCStr {
+    data: SmallVec<[u8; SIZE]>,
```

Comment:
Ah! Now I see what you're doing. Why is this an improvement over the `OnHeap` version?

Reply:
Absent the union optimization, this should generate the same code. However *with* the union optimization, the type is reduced to `(usize, [T; _] / *mut T)`, which saves the discriminant + padding.

Apart from the memory savings, the code has seen a good deal of perf work already.